src/python/pants/engine/scheduler.py

```diff
@@ -122,86 +120,75 @@ def _to_ids_buf(self, types):
-          elif selector_type is SelectVariant:
-            key_buf = self._to_utf8_buf(selector.variant_key)
-            self._native.lib.task_add_select_variant(self._scheduler,
-                                                     product_constraint,
-                                                     key_buf)
-          elif selector_type is SelectLiteral:
-            # NB: Intentionally ignores subject parameter to provide a literal subject.
-            self._native.lib.task_add_select_literal(self._scheduler,
-                                                     self._to_key(selector.subject),
-                                                     product_constraint)
-          elif selector_type is SelectDependencies:
-            self._native.lib.task_add_select_dependencies(self._scheduler,
-                                                          product_constraint,
-                                                          self._to_constraint(selector.dep_product),
-                                                          self._to_utf8_buf(selector.field),
-                                                          self._to_ids_buf(selector.field_types))
-          elif selector_type is SelectTransitive:
-            self._native.lib.task_add_select_transitive(self._scheduler,
-                                                        product_constraint,
-                                                        self._to_constraint(selector.dep_product),
-                                                        self._to_utf8_buf(selector.field),
-                                                        self._to_ids_buf(selector.field_types))
-          elif selector_type is SelectProjection:
-            if len(selector.fields) != 1:
-              raise ValueError("TODO: remove support for projecting multiple fields at once.")
-            field = selector.fields[0]
-            self._native.lib.task_add_select_projection(self._scheduler,
-                                                        self._to_constraint(selector.product),
-                                                        TypeId(self._to_id(selector.projected_subject)),
-                                                        self._to_utf8_buf(field),
-                                                        self._to_constraint(selector.input_product))
-          else:
-            raise ValueError('Unrecognized Selector type: {}'.format(selector))
-        self._native.lib.task_end(self._scheduler)
+        if type(rule) is SingletonRule:
+          self._register_singleton(output_constraint, rule)
+        elif type(rule) is TaskRule:
+          self._register_task(output_constraint, rule)
+        else:
+          raise ValueError('Unexpected Rule type: {}'.format(rule))
+
+  def _register_singleton(self, output_constraint, rule):
+    """Register the given SingletonRule.
+
+    Singleton tasks are those that are the default for a particular type(product).
```

Comment:
I thought that singletons are the only allowed value for that type since we don't allow merges from multiple producers. So, maybe "Singleton tasks are the only task that can produce a product for a particular type(product)"?